GPU/GPUTracking/Global/GPUChainTracking.cxx

```diff
@@ -897,8 +922,30 @@ int GPUChainTracking::RunTPCClusterizer()
       }
 
       if (mIOPtrs.tpcZS) {
-        int firstHBF = mIOPtrs.tpcZS->slice[0].count[0] ? o2::raw::RDHUtils::getHeartBeatOrbit(*(const o2::header::RAWDataHeader*)mIOPtrs.tpcZS->slice[0].zsPtr[0][0]) : 0;
-        runKernel<GPUTPCCFDecodeZS, GPUTPCCFDecodeZS::decodeZS>({doGPU ? clusterer.mPmemory->counters.nPages : GPUTrackingInOutZS::NENDPOINTS, CFDecodeThreadCount(), lane}, {iSlice}, {}, firstHBF);
+        int firstHBF = mIOPtrs.tpcZS->slice[iSlice].count[0] ? o2::raw::RDHUtils::getHeartBeatOrbit(*(const o2::header::RAWDataHeader*)mIOPtrs.tpcZS->slice[iSlice].zsPtr[0][0]) : 0;
+        // clusterer.mPmemory->counters.nDigits = TPCClusterizerDecodeZSCount(iSlice, 1000, 1400);
```

Comment:
@fweig : This PR changes the TPC ZS decoder such that in can decode subslices in time.
What you basically need to to is to uncomment these 2 lines here, and put in the time bin range you want to decode, in the example 1000 - 1400 (inclusive).
It will decode all 8kb pages that overlap with the range you provide, and store the digits consecutively in the array.
clusterer.mPmemory->counters.nDigits is set to the number of digits just decoded.
The decoder will also decode digits outside the provided time range from overlapping 8kb pages.
I cannot easily avoid this, because I don't know how many digits inside a page will fit in the range without decoding it first. And I am precomputing all output offsets. What you can do of course is to pass min and max to the decoder kernel and just write some invalid charge or so, but this can also be filtered out at a later stage.
If relevant, a single page can contain at maximum 256 time bins, so the output can be at maximum 255 time bins outside the range you provide.

Reply:
Perfect, that should be exactly what i need. Ideally i wouldn't wanna decode the timebins outside the range at all. As this would reduce the size of a sub time slice by 510 timebins in the worstcase. But this is only possible by adding braching in the decoder.

It might be useful `DecodeZSCount` if decodeZS also returns the first / last timebin of the overlapping 8kb pages. But i hope i can do that on my own.

Reply:
@fweig : I do not understand what you mean by "It might be useful DecodeZSCount if decodeZS also returns the first / last timebin of the overlapping 8kb pages. But i hope i can do that on my own." You mean you want to get the range of decoded clusters time bins?

Please note also that there might only be 1 overlapping page at both borders, so there will not be many clusters decoded out of the range you provide. I would add a min and max parameter to the decode kernel, and then I would not write charges outside of this range to the chargemap.
Then, you just need some way to filter out the entries in the index map as well.

Reply:
That seems like the easiest option. My idea was to not decode the overlapping 8kb page at the upper border. In that case I'd need the last timebin that will actually be decoded to know where the next sub slice would start. Getting rid of the overlapping page at lower border could be done by also adjusting the start time of the sub slice. That would get rid of range checks in the decoder. But also one less 8kb page per sub slice would be decoded.

Reply:
Not sure if that would work. A time bin might also be partially decoded, i.e. it will be split over two pages, only one will be decoded.